mqtt/mqtt-broker/src/session/state/mod.rs

```diff
@@ -226,36 +224,30 @@ impl SessionState {
         self.try_publish()
     }
 
-    pub fn handle_puback0(
-        &mut self,
-        id: proto::PacketIdentifier,
-    ) -> Result<Option<ClientEvent>, Error> {
-        debug!("discarding QoS 0 packet identifier {}", id);
-        self.waiting_to_be_acked_qos0.remove(&id);
-        self.packet_identifiers_qos0.discard(id);
-        self.try_publish()
-    }
-
-    fn try_publish(&mut self) -> Result<Option<ClientEvent>, Error> {
-        if self.allowed_to_send() {
+    pub(super) fn try_publish(&mut self) -> Result<Option<ClientEvent>, Error> {
+        if let Some(permit) = self.allowed_to_send() {
             if let Some(publication) = self.waiting_to_be_sent.dequeue() {
-                let event = self.prepare_to_send(&publication)?;
+                let event = self.prepare_to_send(&publication, permit)?;
                 return Ok(Some(event));
             }
         }
         Ok(None)
     }
 
-    pub(super) fn allowed_to_send(&self) -> bool {
-        match self.config.max_inflight_messages() {
-            Some(limit) => {
-                let num_inflight = self.waiting_to_be_acked.len()
-                    + self.waiting_to_be_acked_qos0.len()
-                    + self.waiting_to_be_completed.len();
-                num_inflight < limit.get()
-            }
-            None => true,
-        }
+    pub(super) fn allowed_to_send(&self) -> Option<PublishPermit> {
+        self.config
```

Comment:
:crying_cat_face: this is unreadable... can we make it simple with `match` please?